### PR TITLE
feat(auth): add idtoken package

### PR DESCRIPTION
**Pull request by codyoss**

### CODE REVIEW

File: auth/idtoken/compute.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (c computeIDTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
+	v := url.Values{}
+	v.Set("audience", c.audience)
+	if c.format != ComputeTokenFormatStandard {
+		v.Set("format", "full")
+	}
+	if c.format == ComputeTokenFormatFullWithLicense {
+		v.Set("licenses", "TRUE")
```
Comment (noahdietz) [inline auth/idtoken/compute.go:58]: out of curiosity, does this value need to be exactly `TRUE` or is it any truthy value? Just find it funny that there is an all-caps boolean string like this haha

File: auth/idtoken/compute.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	urlSuffix := identitySuffix + "?" + v.Encode()
+	res, err := c.client.Get(urlSuffix)
+	if err != nil {
+		return nil, err
+	}
+	if res == "" {
+		return nil, fmt.Errorf("idtoken: invalid response from metadata service")
```
Comment (noahdietz) [inline auth/idtoken/compute.go:66]: Nit: maybe slightly more specific error? 
```suggestion
		return nil, fmt.Errorf("idtoken: invalid empty response from metadata service")
```

File: auth/idtoken/compute.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	return &auth.Token{
+		Value: res,
+		Type:  internal.TokenTypeBearer,
+		// Compute tokens are valid for one hour
```
Comment (noahdietz) [inline auth/idtoken/compute.go:71]: By chance, is this documented somewhere? If it is, would it make sense to put a link to that doc in this comment? A fair answer is "Yes it is, and no it's not necessary" :)

File: auth/idtoken/file.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		config := impersonate.IDTokenOptions{
+			Audience:        opts.Audience,
+			TargetPrincipal: account,
+			IncludeEmail:    true,
+			Client:          opts.client(),
+			TokenProvider:   creds,
+		}
+		return impersonate.NewIDTokenProvider(&config)
+	default:
+		return nil, fmt.Errorf("idtoken: unsupported credentials type")
```
Comment (noahdietz) [inline auth/idtoken/file.go:108]: Nit: Is it worth putting said type in the error message? Not sure if that is leaky or not.
```suggestion
		return nil, fmt.Errorf("idtoken: unsupported credentials type: %v", t)
```

**Review by noahdietz — APPROVED** [review]
A couple of nits but looking great

File: auth/idtoken/compute.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+type computeIDTokenProvider struct {
+	audience string
+	format   ComputeTokenFormat
+	client   metadata.Client
+}
+
+func (c computeIDTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
+	v := url.Values{}
+	v.Set("audience", c.audience)
+	if c.format != ComputeTokenFormatStandard {
```
Comment (quartzmo) [inline auth/idtoken/compute.go:54]: Did you want to add a unit test case for format == `ComputeTokenFormatStandard`?

File: auth/idtoken/file.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return nil, err
+	}
+	switch t {
+	case internaldetect.ServiceAccountKey:
+		f, err := internaldetect.ParseServiceAccount(b)
+		if err != nil {
+			return nil, err
+		}
+		opts2LO := &auth.Options2LO{
```
Comment (quartzmo) [inline auth/idtoken/file.go:52]: I really like the explicit options type for "2LO" here, this is much easier to follow than the old design.

File: auth/idtoken/idtoken.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	if b := jsonBytes(opts); b != nil {
+		return tokenProviderFromBytes(b, opts)
+	}
+	if metadata.OnGCE() {
+		return computeTokenProvider(opts)
+	}
+	return nil, fmt.Errorf("idtoken: couldn't find any credentials")
+}
+
+func jsonBytes(opts *Options) []byte {
```
Comment (quartzmo) [inline auth/idtoken/idtoken.go:98]: Why not make this a method on `Options`? I believe the `opts != nil` check on 99 is unnecessary after the nil check on 83.

File: auth/idtoken/idtoken.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	ComputeTokenFormatFull
+	// ComputeTokenFormatFullWithLicense means the same as
+	// [ComputeTokenFormatFull] with the addition of claims about licenses
+	// associated with the instance.
+	ComputeTokenFormatFullWithLicense
+)
+
+// Options for the configuration of creation of an ID token with
+// [NewTokenProvider].
+type Options struct {
```
Comment (quartzmo) [inline auth/idtoken/idtoken.go:49]: I love the new dedicated `Options` struct as opposed to the old design which used shared client Options.

File: auth/idtoken/compute.go (hunk 3 above)
Comment (quartzmo) [reply auth/idtoken/compute.go:71]: Note also that this is a change from the old design, which was 55 minutes in order to provide a 5 minute "buffer".

Comment (codyoss) [reply auth/idtoken/compute.go]: added. Yeah removed the buffer as one is aleady builtin by wrapping it with a cachedtokenprovider. Updated the refresher code above to match this just to be safe. Good call out.

File: auth/idtoken/file.go (hunk 6 above)
Comment (codyoss) [reply auth/idtoken/file.go:52]: :)

File: auth/idtoken/idtoken.go (hunk 8 above)
Comment (codyoss) [reply auth/idtoken/idtoken.go:49]: :)

File: auth/idtoken/idtoken.go (hunk 7 above)
Comment (codyoss) [reply auth/idtoken/idtoken.go]: done and updated.

**Review by noahdietz — APPROVED** [review]

**Review by quartzmo — APPROVED** [review]

**Pull request merged**